crates/viewer/re_renderer/src/renderer/video.rs

```diff
@@ -0,0 +1,526 @@
+        media_type: Option<&str>,
+        data: &[u8],
+    ) -> Result<Self, VideoError> {
+        let data = match media_type {
+            Some("video/mp4") => re_video::load_mp4(data)?,
+            _ => return Err(VideoError::Load(VideoLoadError::UnknownMediaType)),
+        };
+        let decoder = VideoDecoder::new(render_context, data).ok_or_else(|| VideoError::Init)?;
+
+        Ok(Self { decoder })
+    }
+
+    pub fn duration_ms(&self) -> f64 {
+        self.decoder.duration_ms()
+    }
+
+    pub fn width(&self) -> u32 {
+        self.decoder.width()
+    }
+
+    pub fn height(&self) -> u32 {
+        self.decoder.height()
+    }
+
+    pub fn get_frame(&mut self, timestamp_s: f64) -> GpuTexture2D {
+        re_tracing::profile_function!();
+        self.decoder.get_frame(TimeMs::new(timestamp_s * 1e3))
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum VideoError {
+    #[error("{0}")]
+    Load(#[from] VideoLoadError),
+    #[error("failed to initialize video decoder")]
+    Init,
+}
+
+use decoder::VideoDecoder;
+
+fn alloc_video_frame_texture(
+    device: &wgpu::Device,
+    pool: &GpuTexturePool,
+    width: u32,
+    height: u32,
+) -> GpuTexture2D {
+    let Some(texture) = GpuTexture2D::new(pool.alloc(
+        device,
+        &TextureDesc {
+            label: "video".into(),
+            size: wgpu::Extent3d {
+                width,
+                height,
+                depth_or_array_layers: 1,
+            },
+            mip_level_count: 1,
+            sample_count: 1,
+            dimension: wgpu::TextureDimension::D2,
+            format: wgpu::TextureFormat::Rgba8Unorm,
+            usage: wgpu::TextureUsages::COPY_DST
+                | wgpu::TextureUsages::TEXTURE_BINDING
+                | wgpu::TextureUsages::RENDER_ATTACHMENT,
+        },
+    )) else {
+        unreachable!();
+    };
+
+    texture
+}
+
+// fn latest_at<'a, T, K: Ord>(v: &'a [T], key: impl Fn(&T) -> K, needle: &K) -> Option<&'a T> {
+//     latest_at_idx(v, key, needle).map(|idx| &v[idx])
+// }
+
+/// Returns the index of:
+/// - The index of `needle` in `v`, if it exists
+/// - The index of the first element in `v` that is lesser than `needle`, if it exists
+/// - `-1`, if `v` is empty OR `needle` is greater than all elements in `v`
+fn latest_at_idx<T, K: Ord>(v: &[T], key: impl Fn(&T) -> K, needle: &K) -> Option<usize> {
+    fn inner<T, K: Ord>(v: &[T], key: impl Fn(&T) -> K, needle: &K) -> isize {
+        let mut left = 0isize;
+        let mut right = v.len() as isize;
+
+        if right == 0 {
+            return -1;
+        }
+
+        if right > 0 && &key(&v[0]) > needle {
+            return -1;
+        }
+
+        while left < right {
+            let middle = (left + right) / 2;
+
+            if &key(&v[middle as usize]) <= needle {
+                left = middle + 1;
+            } else {
+                right = middle;
+            }
+        }
+
+        left - 1
+    }
```

Comment:
You could use https://doc.rust-lang.org/stable/std/primitive.slice.html#method.partition_point here